### PR TITLE
Persistance

### PR DESCRIPTION
**Pull request by krl**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+Change the name of `BranchMutMappedMut` to `MappedBranchMut`
```
Comment (ZER0) [inline CHANGELOG.md:11]: Add `- ` at the beginning of the list item (see https://keepachangelog.com/en/1.0.0/)

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
+### Removed
+Remove `path` from branch traversal, case now covered by `walk`
```
Comment (ZER0) [inline CHANGELOG.md:14]: Same here.

File: src/lib.rs [hunk 3]
```diff
@@ -13,8 +13,8 @@
 //! `Branch` and `BranchMut`, types for representing branches in tree-formed
 //! data as well as methods of search.
 
-#![no_std]
-#![deny(missing_docs)]
+#![cfg_attr(not(feature = "persistance"), no_std)]
+#![warn(missing_docs)]
```
Comment (ZER0) [inline src/lib.rs:17]: Why the change from `deny` to `warn`? I assume it's a left over did during the development, it should be change back.

File: src/persist/disk.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    where
+        P: Into<PathBuf>,
+    {
+        let path = path.into();
+
+        let mut index_path = path.clone();
+        let mut data_path = path.clone();
+
+        index_path.push("index");
+        data_path.push("data");
```
Comment (ZER0) [inline src/persist/disk.rs:46]: Use `join` instead:
```rust
let index_path = path.join("index");
let data_path = path.join("data");
```

File: src/persist/disk.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    fn get(&self, id: &Id) -> Result<GenericTree, PersistError> {
+        if let Some(ofs) = self.index.get(id)? {
+            let mut data = File::open(&self.data_path)?;
+
+            data.seek(SeekFrom::Start(*ofs))?;
+
+            let len = id.size();
+
+            let mut buf = Vec::with_capacity(len);
+            buf.resize_with(len, || 0);
```
Comment (ZER0) [inline src/persist/disk.rs:79]: You can just use `let mut buf = vec![0; len];`

File: src/persist/mod.rs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    /// Returns a constructor for a temporary test backend
+    pub fn test_backend_ctor() -> BackendCtor<TestBackend> {
+        BackendCtor::new(|| {
+            let dir = tempfile::tempdir().unwrap();
+            let b = DiskBackend::new(dir.path()).unwrap();
+            TestBackend(b, dir)
+        })
+    }
```
Comment (ZER0) [inline src/persist/mod.rs:167]: Test method shouldn't be part of the API, we should have it as helper function in the integration test.

File: src/persist/test.rs [hunk 7]
```diff
@@ -0,0 +1,32 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
```
Comment (ZER0) [inline src/persist/test.rs:6]: Why this file is inside `src`?
This should be part of the integration test.

File: src/persist/test.rs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+use tempfile::TempDir;
+
+use crate::{Backend, DiskBackend, GenericTree, PersistError, PutResult};
+use canonical::Id;
+
+pub struct TestBackend(pub DiskBackend, pub TempDir);
+
+impl TestBackend {
+    pub fn new() -> Self {
+        todo!()
```
Comment (ZER0) [inline src/persist/test.rs:16]: Is that a leftover? Otherwise you should explain something in this `todo`.

File: Cargo.toml [hunk 9]
```diff
@@ -9,9 +9,17 @@ license = "MPL-2.0"
...
 [dependencies]
-canonical = "0.6"
-canonical_derive = "0.6"
+appendix = { version = "0.2.0", optional = true }
+parking_lot = { version = "0.11.1", optional = true }
+lazy_static = { version = "1.4.0", optional = true }
+canonical = "0.6.3"
+canonical_derive = "0.6.0"
+arbitrary = { version = "1.0.1", features = ["derive"] }
+tempfile = { version = "3.2.0", optional = true }
```
Comment (ZER0) [inline Cargo.toml:18]: This should be only a dep for tests (see below).

**Review by ZER0 — CHANGES_REQUESTED** [review]
LGTM, but there are few things that I believe are either leftovers or need to be addressed differently. The main concern is about the test API that is part of the actual API, so before landing and releasing we should address that as well.

An additional couple of things:
1. Change the name of the base target since it's not `0.7.2` anymore but it's `0.8.0` and that could lead to confusion.
2. Remove from the github's workflow any reference to the toolchain and the override (as we did in other repo) since you have the `rust-toolchain` file. And set the `rust-toolchain` to the latest nighly that supports also `rustfmt` (I think in the other repo we have `2021-06-06`).
This is unfortunate but we notice that having in the override in the CI only for the rustfmt was leading to messy configuration and heard to keep consistent across repos. In this way at least each repo is responsible for itself and the CI and the local development is consistent, and it's possible to update the repo's rust-toolchain without the needs to update the others.

File: src/annotations/max_key.rs [hunk 10]
```diff
@@ -81,26 +81,27 @@ where
...
+        C: Compound<A>,
+        A: Annotation<C::Leaf>,
+    {
         let mut max = MaxKey::NegativeInfinity;
 
-        for child in node.children() {
-            let ann = &*child.annotation();
-            let m = ann.borrow();
+        for ann in iter {
+            let m = (*ann).borrow();
```
[the remark this answers is not in the document]
Comment (krl) [reply src/annotations/max_key.rs:104]: I changed it to a fold, i'm not seeing how max_by would help here. Note that we need to avoid cloning if unneccesary.

File: src/persist/disk.rs (hunk 5 above)
Comment (krl) [reply src/persist/disk.rs:79]: true.

File: src/persist/test.rs (hunk 7 above)
Comment (krl) [reply src/persist/test.rs]: leftover as you assumed

File: src/lib.rs (hunk 3 above)
Comment (krl) [reply src/lib.rs]: fixed

File: .github/workflows/dusk_ci.yml [hunk 11]
```diff
@@ -41,7 +42,6 @@ jobs:
       - uses: actions-rs/toolchain@v1
         with:
           profile: minimal
-          toolchain: nightly-2020-10-25
           override: true
```
Comment (ZER0) [inline .github/workflows/dusk_ci.yml:45]: Nit: this also needs to be removed (see again: https://github.com/dusk-network/rusk/pull/320/files as reference), you don't have anything to override anymore.

File: CHANGELOG.md [hunk 12]
```diff
@@ -7,11 +7,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+### Removed
+- Remove `path` from branch traversal, case now covered by `walk`
+
 ## [0.7.2] - 2021-06-01
 
 ### Added
-Add `IntoIterator` to `MappedBranch` and `BranchMutMappedMut`
-Add support for non-mutably dereferencing mutable mapped branches
+- Add `IntoIterator` to `MappedBranch` and `BranchMutMappedMut`
+- Add support for non-mutably dereferencing mutable mapped branches
```
Comment (ZER0) [inline CHANGELOG.md:20]: I'm not quite sure: what changed here?

File: rust-toolchain [hunk 13]
```diff
@@ -1 +1 @@
-nightly
+nightly-2021-06-06
```
Comment (ZER0) [inline rust-toolchain:1]: Add a new line at the end of the file

**Review by ZER0 — APPROVED** [review]
LGTM Just small nits to fixed before the merge, see the comments, but no needs for another round of review if fixed.
